### PR TITLE
Remove imagery provider information from BackgroundMapSettings

### PR DESCRIPTION
**Pull request by pmconne**
See NextVersion.md.

BackgroundMapSettings has a provider name and map type for base background map imagery. Options are from a fixed list (bing or mapbox, 3 types of imagery).
MapImagerySettings has a "background base" layer that defines base background map imagery (or solid color). Options are any type of imagery from any source.
Today, nearly everybody who wants to change the imagery does so using Viewport.changeBackgroundMapProps, passing in a BackgroundMapProps that defines all the properties (including provider/type) they want to change; the rest retain their current values.
This requires synchronization between MapImagerySettings and BackgroundMapSettings and leads to redundant storage of the base imagery information.

The new, correct way to define the base imagery layer is via MapImagerySettings.
BackgroundMapSettings no longer exposes a provider name or map type property. The corresponding JSON properties are deprecated; they are intended for use only at the persistence layer. BackgroundMapSettings does preserve the imagery information internally so that it is preserved in its JSON representation.
For backwards compatibility, when we instantiate a DisplayStyleSettings from its persistent JSON representation, if no base imagery layer is specified we use the imagery specified by those deprecated BackgroundMapProps properties.

The type system allows us to turn accidental/incorrect usage of the deprecated properties into compile-time errors:
- BackgroundMapProps is the in-memory JSON representation. It declares providerName and providerData as type `never`; trying to set them produces a type error.
- DeprecatedBackgroundMapProps holds (only) the deprecated properties.
- DisplayStyleSettingsProps.backgroundMap is of the type PersistentBackgroundMapProps, which includes all of BackgroundMapProps and DeprecatedBackgroundMapProps.
- BackgroundMapSettings now has to/fromJSON returning BackgroundMapProps and to/fromPersistentJSON returning PersistentBackgroundMapProps.

Usage of DisplayStyleState.changeBackgroundMapProps to change the imagery can be replaced with DisplayStyleState.changeBackgroundMapProvider.

I imagine this kind of approach may become more common as the iModel schema evolves and we deprecate more persistence formats. For example, if in the future we deprecate more BackgroundMapProps properties, they can be moved to DeprecatedBackgroundMapProps and the compiler should catch any mistakes (barring blatant subversion of the type system or use of raw Javascript).

TODO in the (hopefully near) future:
- Clean up MapImagerySettings and MapLayerSettings APIs.
  - Arrays should be replaced with a sub-type of ReadonlyArray that permits mutation through specific methods, all of which raise events.
- Make sure DisplayStyleSettings is notified (and dispatches notifications) when MapImagery/LayerSettings change - via the events mentioned above.
  - Currently you must use DisplayStyleState APIs which call an internal synchMapImagery function which updates the DisplayStyleSettings' json and dispatch the event - using the API in imodeljs-common causes JSON to get out of sync and does not dispatch events.
- Clean up DisplayStyleState's constellation of APIs for changing map layer settings, which appear to have been designed primarily for ease of implementing tools for testing the map layers code during its development. They primarily consist of finding a layer's index by name, url, and overlay/background; then passing that index to another API that can update/delete/insert it. This is not very ergonomic or intelligible for someone trying to use the API more directly.
  - Most if not all of these APIs should be replaced with cleaner APIs on MapImagery/LayerSettings and should not be index-based.
- The fromJSON methods of MapLayerSettings, MapSubLayerSettings, and BaseMapLayerSettings should never return undefined. They should require caller to pass a valid JSON object with `url` and `name` defined (or whatever else they require), and enforce that through the type system at compile time.

### CODE REVIEW

File: core/common/src/MapImagerySettings.ts [hunk 1]
```diff
@@ -31,7 +31,17 @@ export interface MapImageryProps {
...
  */
-export type BaseLayerSettings = MapLayerSettings | ColorDef;
+export type BaseLayerSettings = BaseMapLayerSettings | ColorDef;
+
+// eslint-disable-next-line @typescript-eslint/no-redeclare
+export namespace BaseLayerSettings {
+  export function fromJSON(props: BaseLayerProps): BaseLayerSettings {
+    const settings = typeof props === "number" ? ColorDef.fromJSON(props) : BaseMapLayerSettings.fromJSON(props);
+
+    // ###TODO handle undefined BaseMapLayerSettings differently (what should be default?), or make MapImagerySettings.fromJSON possibly return undefined.
```
Comment (pmconne) [inline core/common/src/MapImagerySettings.ts:41]: @mdastous-bentley ?

Comment (mdastous-bentley) [reply core/common/src/MapImagerySettings.ts:41]: I asssume if someone turned ON ViewFlags.backgroundMap, he would expect to see a map appears even though he didnt customized the map settings? In that case I guess I guess it should be bing/hybrid.

BackgroundMapProvider.fromJSON and BackgroundMapSettings._provider also defaults to bing/hybrid

File: core/common/src/MapLayerSettings.ts [hunk 2]
```diff
@@ -380,3 +346,112 @@ export class MapLayerSettings {
...
+    if (this.provider)
+      props.provider = this.provider.toJSON();
+
+    return props;
+  }
+
+  public override clone(changedProps: MapLayerProps): BaseMapLayerSettings {
+    const prevUrl = this.url;
+    const clone = super.clone(changedProps);
+    assert(clone instanceof BaseMapLayerSettings);
```
Comment (mdastous-bentley) [inline core/common/src/MapLayerSettings.ts:382]: Im seeing assertion failures here, clone appears to be an instance of MapLayerSettings and not BaseMapLayerSettings

Comment (pmconne) [reply core/common/src/MapLayerSettings.ts:382]: I have fixed that already. I am working on these 3 full stack tests.

**Comment by pmconne** [comment]
> DisplayTestApp freeze and MapTileTree.getLayerTransparency gives an assert

I will take a look after I finish documentation, unless you want to investigate meanwhile.

File: core/common/src/MapLayerSettings.ts [hunk 3]
```diff
@@ -380,3 +347,140 @@ export class MapLayerSettings {
...
+ * @beta
+ */
+export class BaseMapLayerSettings extends MapLayerSettings {
+  private _provider?: BackgroundMapProvider;
+
+  /** The provider from which this base layer was configured, if any. */
+  public get provider(): BackgroundMapProvider | undefined { return this._provider; }
+
+  /** Create a base layer from its JSON representation.
+   * TODO: This, MapLayerSettings.fromJSON, and MapSubLayerSettings.fromJSON should never return undefined.
```
Comment (pmconne) [inline core/common/src/MapLayerSettings.ts:373]: @mdastous-bentley probably a separate PR?

Comment (mdastous-bentley) [reply core/common/src/MapLayerSettings.ts:373]: So make MapLayerSettings.fromJSON/ BaseMapLayerSettings.fromJSON defaults to bing right?

Im unsure what aould be proper defaults for MapSubLayerSettings.fromJSON, we might end up with subLayers with empty names.  I guess MapImagerySettings.ctor can check if name if empty, dont add itto _backgroundLayers/_overlayLayers.  What do you think?

Yes that can made part of another PR.

Comment (pmconne) [reply core/common/src/MapLayerSettings.ts:373]: No, I'm not suggesting we produce a default if we can't make a layer out of the inputs. I'm suggesting that, for example, MapLayerSettings.fromJSON takes an argument that meets the minimum requirements `{ url: string, name: string }` (neither can be undefined) such that you never bail like you currently do:
```
  public static fromJSON(json?: MapLayerProps): MapLayerSettings | undefined {
    if (json === undefined || json.url === undefined || undefined === json.name)
      return undefined;
```

**Review by bbastings — APPROVED** [review]

**Comment by pmconne** [comment]
> I will take a look after I finish documentation, unless you want to investigate meanwhile.

I pushed the fix.

**Review by markschlosseratbentley — APPROVED** [review]

**Review by RBBentley — APPROVED** [review]

**Review by mdastous-bentley — APPROVED** [review]

**Comment by pmconne** [comment]
@Mergifyio update

File: core/common/src/MapLayerSettings.ts [hunk 4]
```diff
@@ -127,9 +136,12 @@ export interface MapLayerProps {
   transparentBackground?: boolean;
   /** Is a base layer.  Defaults to 'false'. */
   isBase?: boolean;
-  /** Access Key for the Layer, like a subscription key or access token */
+  /** Access Key for the Layer, like a subscription key or access token.
+   * TODO This does not belong in the props object. It should never be persisted.
```
Comment (pmconne) [inline core/common/src/MapLayerSettings.ts:140]: @mdastous-bentley I changed toJSON to exclude this property, but it really doesn't belong in MapLayerProps at all. Revisit in future PR please?

**Pull request merged**